File: backends/spring-boot/src/main/java/com/hellotimepro/springboot/web/AuthController.java
package com.hellotimepro.springboot.web;

import com.hellotimepro.springboot.dto.Dtos.AuthTokens;
import com.hellotimepro.springboot.dto.Dtos.Envelope;
import com.hellotimepro.springboot.dto.Dtos.LoginRequest;
import com.hellotimepro.springboot.dto.Dtos.LogoutRequest;
import com.hellotimepro.springboot.dto.Dtos.RefreshRequest;
import com.hellotimepro.springboot.dto.Dtos.RegisterRequest;
import com.hellotimepro.springboot.service.AuthService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/v1/auth")
public class AuthController {
  private final AuthService auth;

  public AuthController(AuthService auth) {
    this.auth = auth;
  }

  @PostMapping("/register")
  @ResponseStatus(HttpStatus.CREATED)
  public Envelope<AuthTokens> register(@Valid @RequestBody RegisterRequest req) {
    return Envelope.ok(auth.register(req));
  }

  @PostMapping("/login")
  public Envelope<AuthTokens> login(@Valid @RequestBody LoginRequest req) {
    return Envelope.ok(auth.login(req));
  }

  @PostMapping("/refresh")
  public Envelope<AuthTokens> refresh(@Valid @RequestBody RefreshRequest req) {
    return Envelope.ok(auth.refresh(req.refreshToken()));
  }

  @PostMapping("/logout")
  public ResponseEntity<Void> logout(@RequestBody(required = false) LogoutRequest req) {
    auth.logout(req == null ? null : req.refreshToken());
    return ResponseEntity.noContent().build();
  }
}

File: backends/aspnet/src/Controllers/AuthController.cs
using HelloTimePro.Aspnet.Dto;
using HelloTimePro.Aspnet.Services;
using HelloTimePro.Aspnet.Web;
using Microsoft.AspNetCore.Mvc;

namespace HelloTimePro.Aspnet.Controllers;

[Route("api/v1/auth")]
public sealed class AuthController : ApiControllerBase
{
    private readonly AuthService _auth;
    public AuthController(AuthService auth) => _auth = auth;

    [HttpPost("register")]
    public async Task<IActionResult> Register([FromBody] RegisterRequest? req)
    {
        var tokens = await _auth.Register(req ?? new RegisterRequest(null, null, null, null));
        return Wrapped(tokens, StatusCodes.Status201Created);
    }

    [HttpPost("login")]
    public async Task<IActionResult> Login([FromBody] LoginRequest? req)
    {
        var tokens = await _auth.Login(req ?? new LoginRequest(null, null));
        return Wrapped(tokens);
    }

    [HttpPost("refresh")]
    public async Task<IActionResult> Refresh([FromBody] RefreshRequest? req)
    {
        var tokens = await _auth.Refresh(req?.RefreshToken);
        return Wrapped(tokens);
    }

    [HttpPost("logout")]
    public async Task<IActionResult> Logout([FromBody] LogoutRequest? req)
    {
        await _auth.Logout(req?.RefreshToken);
        return StatusCode(StatusCodes.Status204NoContent);
    }
}
